hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -220,25 +158,171 @@ protected List<DatanodeDetails> chooseDatanodesInternal(
+          final List<DatanodeDetails> excludedNodes,
+          final List<DatanodeDetails> favoredNodes,
+          final int nodesRequiredToChoose, final long metadataSizeRequired,
+          final long dataSizeRequired) throws SCMException {
+    if (nodesRequiredToChoose <= 0) {
+      String errorMsg = "num of nodes required to choose should bigger" +
+          "than 0, but the given num is " + nodesRequiredToChoose;
       throw new SCMException(errorMsg, null);
     }
+    metrics.incrDatanodeRequestCount(nodesRequiredToChoose);
+    int nodesRequired = nodesRequiredToChoose;
+    int excludedNodesCount = excludedNodes == null ? 0 : excludedNodes.size();
+    List<Node> availableNodes = networkTopology.getNodes(
+        networkTopology.getMaxLevel());
+    int totalNodesCount = availableNodes.size();
+    if (excludedNodes != null) {
+      availableNodes.removeAll(excludedNodes);
+    }
+    if (availableNodes.size() < nodesRequired) {
+      throw new SCMException("No enough datanodes to choose. " +
+          "TotalNode = " + totalNodesCount +
+          " AvailableNode = " + availableNodes.size() +
+          " RequiredNode = " + nodesRequired +
+          " ExcludedNode = " + excludedNodesCount,
+          SCMException.ResultCodes.FAILED_TO_FIND_SUITABLE_NODE);
+    }
+
+    List<DatanodeDetails> mutableFavoredNodes = new ArrayList<>();
+    if (favoredNodes != null) {
+      // Generate mutableFavoredNodes, only stores valid favoredNodes
+      for (DatanodeDetails datanodeDetails : favoredNodes) {
+        if (isValidNode(datanodeDetails, metadataSizeRequired,
+            dataSizeRequired)) {
+          mutableFavoredNodes.add(datanodeDetails);
+        }
+      }
+      Collections.shuffle(mutableFavoredNodes);
+    }
+    if (excludedNodes != null) {
+      mutableFavoredNodes.removeAll(excludedNodes);
+    }
+
+    if (usedNodes == null) {
+      usedNodes = Collections.emptyList();
+    }
+    List<Node> racks = getAllRacks();
+    Set<Node> usedRacks = usedNodes.stream()
+            .map(node -> networkTopology.getAncestor(node, RACK_LEVEL))
+            .filter(node -> node != null)
+            .collect(Collectors.toSet());
+    int requiredReplicationFactor = usedNodes.size() + nodesRequired;
+    int numberOfRacksRequired =
+            getRequiredRackCount(requiredReplicationFactor);
+    int additionalRacksRequired = numberOfRacksRequired - usedRacks.size();
+    if (nodesRequired < additionalRacksRequired) {
+      String reason = "Required nodes size: " + nodesRequired
+              + " is less than required number of racks to choose: "
+              + additionalRacksRequired + ".";
+      LOG.warn("Placement policy cannot choose the enough racks. {}"
+                      + "Total number of Required Racks: {} Used Racks Count:" +
+                      " {}, Required Nodes count: {}",
+              reason, numberOfRacksRequired, usedRacks.size(), nodesRequired);
+      throw new SCMException(reason,
```

Comment:
There are a few new `throw new SCMException(reason,` here I think - are we making this class more strict, in that if it cannot spread the nodes across racks it will throw, or are we keeping it largely as it was?

Reply:
We are keeping the logic as it is, that it should spread across as many racks as possible, based on the usedNodes & the nodes selected.